bcdb: authenticate request to the API

zaibon opened the pull request.

zaibon committed:
bcdb: authenticate request to the API

muhamadazmy commented on tools/bcdb_mock/mw/auth.go:
```diff
@@ -0,0 +1,103 @@
+
+	f = f.WithID(schema.ID(uid))
+
+	col := m.db.Collection(types.UserCollection)
```
Why not `f.Get()` ?

muhamadazmy commented on tools/bcdb_mock/mw/auth.go:
```diff
@@ -0,0 +1,103 @@
+	db *mongo.Database
+}
+
+func (m mongoKeyGetter) GetKey(id string) interface{} {
```
Shouldn't this be different for different endpoints. for example, the node endpoints where the node initialize the code should check against the Node public key, not the user public key

zaibon replied:
You are correct.

muhamadazmy commented on tools/bcdb_mock/mw/auth.go:
```diff
@@ -0,0 +1,103 @@
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		err := v.Verify(req)
+		if err != nil {
+			w.Header()["WWW-Authenticate"] = []string{challenge}
```
I don't get this actually, does this mean that the user (or the node) has to do 2 trips to execute an action. First one to get Unauthorized + the header to sign, then a request with the correct signature ?

zaibon replied:
Well if you send a request with the proper header right away then you will not get the 401 error.
Here we just notify the client that if he wants to authenticate, it needs to use these header to create the signature.

muhamadazmy replied:
Yes, I understand but if the client will not know what are the required headers unless at least one call is made. right? also does the challenge change with every request?

zaibon replied:
The challenge doesn't not change with every request no. And this needs to be documented. Normal user will always include this during each requests.

zaibon committed:
load private key in BCDB client in all daemon

zaibon committed:
Limit usage of some endpoints to only the authorized users

- Most of the nodes endpoint can only be called by the node itself
`ConfigurePublic` can only be called by the farmer of the node

- Added delete method for both farm and nodes

muhamadazmy commented on tools/bcdb_mock/pkg/directory/farms_handlers.go:
```diff
@@ -90,3 +91,36 @@ func (s *FarmAPI) getFarm(r *http.Request) (interface{}, mw.Response) {
 	return farm, nil
 }
+
+func (s *FarmAPI) delete(r *http.Request) (interface{}, mw.Response) {
```
Why do we need that?
I mean in block chain database, the "delete" should not be allowed (or at least the data should not get deleted from the database, may be only marked as deleted)

zaibon replied:
You are right I should not delete anything but use another mechanism to marked them as deleted. I'll remove from this PR and open an issue to discuss how to handle clean up in a better way.

muhamadazmy commented on tools/bcdb_mock/pkg/directory/node_handlers.go:
```diff
@@ -185,21 +236,70 @@ func (s *NodeAPI) updateUptimeHandler(r *http.Request) (interface{}, mw.Response
 	return nil, nil
 }
+
+func (s *NodeAPI) delete(r *http.Request) (interface{}, mw.Response) {
```
Same question regarding farme.delete()

muhamadazmy reviewed (approved):
This looks really good, I approved but I had some questions that are not blocking

zaibon committed:
explorer: remove delete methods from node and farm API

zaibon committed:
address linting warnings
